### PR TITLE
Plans 2023: pass free-plan action overrides as prop + fix for domain upsell flow

### PR DESCRIPTION
**Pull request by chriskmnds**
Related to #76442

## Proposed Changes

Tackles one of the TODOs in https://github.com/Automattic/wp-calypso/pull/76442 -> https://github.com/Automattic/wp-calypso/pull/76442/files#r1182881397

- Add a bare provision for overriding some of the actions in the grid (which can be extended for overriding more actions if need be). This is done via a `PlanActionOverrides` optional prop.
- Moves the consuming end of the plans 2023 grid into a rough functional component inside `PlanFeaturesMain`. It's a minimal change to just get the wheels in motion. Something to be typed and extracted into its own file at a later time.
- Fixes the domain upsell dialog that's triggered via the `get_domain` query arg. This appears to have been broken for some time (or otherwise not cleaned up if for some reason it was decided as obsolete/unused). 🤔 

### Media

**No plan action override (paid site):**

Screenshot 2023-05-03 at 2 37 07 PM

**Free plan action override (free site):**

Screenshot 2023-05-03 at 2 36 04 PM

**Free plan action override and domain upsell (`get_domain` query arg):**

Screenshot 2023-05-03 at 2 36 36 PM

## Testing Instructions

* Go to `/plans/[site]` and `/start/plans` with a free plan and ensure the free plan action renders as expected
* For `/plans/[site]`, the free plan button should be:
    * `contact support` when current site is not on a free plan
    * `manage add-ons` when current site is on free plan -> should redirect to add-ons section
    * `keep my plan` when `?get_domain=foo.bar` query arg is added and current site is on free plan -> clicking on "keep my plan" should trigger the modal shown above

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/my-sites/plans-features-main/index.jsx
Comment (chriskmnds) [inline client/my-sites/plans-features-main/index.jsx]: This is fine to live here for now AFAIC. We'll extract and type properly at a later point.

File: client/my-sites/plan-features-2023-grid/components/actions.tsx
Comment (aneeshd16) [inline client/my-sites/plan-features-2023-grid/components/actions.tsx]: Probably missed removing this :)

**Review by aneeshd16 — COMMENTED** [review]
Code wise looks good, and works well in my initial testing.

I'll do another round of testing once the merge conflicts are resolved. 👍

File: client/my-sites/plan-features-2023-grid/types.ts
Comment (ddc22) [inline client/my-sites/plan-features-2023-grid/types.ts]: Is there reason why we go with a nested complex data structure than a simpler key value, dictionary 🤔

Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/types.ts]: What exactly do you have in mind? I'm trying to guess from your comment. "simpler key value, dictionary" could be many things

Something flatter you mean i.e. `PlanActionOverrides.loggedInFreePlan`, or even flatter `loggedInFreePlanActionText`, etc?

I'm ok with any option, as long as it's a little DRY. These feel very much temporary to get us to a package. I think the grid is context-aware more than it needs to be with the actions, and we ought to come back to revisit at some point. There is a [follow-up issue](https://github.com/Automattic/wp-calypso/issues/76662) that I created, but for now, I think we patch in whatever way feels more natural.

Comment (ddc22) [reply client/my-sites/plan-features-2023-grid/types.ts]: I meant something like `PlanActionOverrides.loggedInFreePlan` 👍 And #76662 sounds perfect!

Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/types.ts]: works for me @jdc91 👍  updated

File: client/my-sites/plan-features-2023-grid/components/actions.tsx
Comment (chriskmnds) [reply client/my-sites/plan-features-2023-grid/components/actions.tsx]: Thanks @aneeshd16 . removed

**Comment by chriskmnds** [comment]
Thanks for the reviews @aneeshd16 @jdc91 🙌 

This is rebased and ready for another round. I think the domain suggestions modal work came after the last rebase, so need to make sure nothing broken from there

**Review by ddc22 — APPROVED** [review]
Works as expected! We are moving quickly towards the complete decoupling 🦾 
LGTM 🚢

**Review by ddc22 — APPROVED** [review]
Works as expected! We are moving quickly towards the complete decoupling 🦾 
LGTM 🚢

**Review by ddc22 — APPROVED** [review]
Works as expected! We are moving quickly towards the complete decoupling 🦾 
LGTM 🚢

**Review by ddc22 — APPROVED** [review]
Works as expected! We are moving quickly towards the complete decoupling 🦾 
LGTM 🚢

**Comment by ddc22** [comment]
For the record I did not approve this 4 time 😆 , Something wrong with Github !

**Comment by chriskmnds** [comment]
> For the record I did not approve this 4 time 😆 , Something wrong with Github !

and I thought you got over excited! 😝

**Pull request merged**
